File: Algorithms/insertion sort/insertion_sort.cpp
#include <iostream>
using namespace std;





void insertion_sort(int array[], int size)
{
    for (int i = 1; i <= size; i++)
        for (int j = i-1; j >= 0; j--)
        {
            if(array[j] > array[j+1])
            {
                int hold = array[j];
                array[j] = array[j+1];
                array[j+1] = hold;
            }

            else
                break;
        }
}




// a simple function for printing arrays 
void printArray(int array[], int size)
{
    cout << "[";

    for (int i = 0; i < size; i++)
        cout << array[i] << ", ";

    cout << "\b\b]\n";
}
 





int main()
{
    cout << "\n\n ---------------- \n\n";
    cout << "a simple test:\n\n";
    

    int array[] = {5, 113, 4, 0, -1, 3, -11, 43, 354, 132, 26, 33};

    cout << "before sort: ";
    printArray(array, 12);


    insertion_sort(array, 12);


    cout << "after sort: ";
    printArray(array, 12);


    cout << "\n ---------------- \n\n\n";



    return 0;
}

File: Algorithms/insertion sort/InsertionSort.java
import java.util.Arrays;




/**
 * Use {@link InsertionSort#insertionSort(int...)} to sort your {@code int} Array
 * 
 * 
 * @author Mohammad Mahdi Malmasi
 * @version 0.1.0
 */
public class InsertionSort
{
    /**
     * This method sorts your int array with insertion sort algorithms
     * 
     * @param array : array to sort
     */
    public static void insertionSort(int...array)
    {
        // required variables
        int hold = 0, j = 0;


        for (int i = 2; i < array.length; i++)
        {
            // hold to compare with others
            hold = array[i];

            // divide list to two parts (array[0] to array[j] and array[j+1] to array[size-1])
            j = i-1;


            // compare 'hold' with other variables in array[0] to array[j] part
            while (j >= 0 && array[j] > hold)
            {
                // shift variable
                array[j+1] = array[j];

                j--;
            }

            array[j+1] = hold;
        }
    }



    /**
     * A simple example for {@link InsertionSort#insertionSort(int...)}
     *  
     * @param args
     */
    public static void main(String[] args) 
    {
        System.out.println("\n\n ---------------- \n");
        System.out.println("a simple test:\n");
        

        int[] array = {8, 10, 252, 0, -34, 23, -11, 43, 50, 101, 7, 33, 12};
        System.out.println("before sort: " + Arrays.toString(array));


        insertionSort(array);
        System.out.println("after sort: " + Arrays.toString(array));


        System.out.println("\n ---------------- \n\n");
    }
}